versioned/tiered/dynamodb/src/main/java/org/projectnessie/versioned/dynamodb/DynamoStore.java

```diff
@@ -201,59 +219,63 @@ public void close() {
-          return KeysAndAttributes.builder().keys(keys).consistentRead(true).build();
-        }));
-
-        BatchGetItemResponse response = client.batchGetItem(BatchGetItemRequest.builder().requestItems(loads).build());
-        Map<String, List<Map<String, AttributeValue>>> responses = response.responses();
-        Sets.SetView<String> missingElements = Sets.difference(loads.keySet(), responses.keySet());
-        Preconditions.checkArgument(missingElements.isEmpty(), "Did not receive any objects for table(s) %s.", missingElements);
-
-        for (String table : responses.keySet()) {
-          List<LoadOp<?>> loadList = l.get(table);
-          List<Map<String, AttributeValue>> values = responses.get(table);
-          int missingResponses = loadList.size() - values.size();
-          if (missingResponses != 0) {
-            ValueType<?> loadType = loadList.get(0).getValueType();
-            if (loadType == ValueType.REF || loadType == ValueType.L1) {
-              throw new NotFoundException("Unable to find requested ref.");
+    try {
+      while (true) { // for each load step in the chain.
+        List<ListMultimap<String, LoadOp<?>>> stepPages = paginateLoads(loadstep, paginationSize);
+
+        for (ListMultimap<String, LoadOp<?>> l : stepPages) {
+          Map<String, KeysAndAttributes> loads = l.keySet().stream().collect(Collectors.toMap(Function.identity(), table -> {
+            List<LoadOp<?>> loadList = l.get(table);
+            List<Map<String, AttributeValue>> keys = loadList.stream()
+                .map(load -> Collections.singletonMap(DynamoBaseValue.ID, idValue(load.getId())))
+                .collect(Collectors.toList());
+            return KeysAndAttributes.builder().keys(keys).consistentRead(true).build();
+          }));
+
+          BatchGetItemResponse response = client.batchGetItem(BatchGetItemRequest.builder().requestItems(loads).build());
+          Map<String, List<Map<String, AttributeValue>>> responses = response.responses();
+          Sets.SetView<String> missingElements = Sets.difference(loads.keySet(), responses.keySet());
+          Preconditions.checkArgument(missingElements.isEmpty(), "Did not receive any objects for table(s) %s.", missingElements);
+
+          for (String table : responses.keySet()) {
+            List<LoadOp<?>> loadList = l.get(table);
+            List<Map<String, AttributeValue>> values = responses.get(table);
+            int missingResponses = loadList.size() - values.size();
+            if (missingResponses != 0) {
+              ValueType<?> loadType = loadList.get(0).getValueType();
+              if (loadType == ValueType.REF || loadType == ValueType.L1) {
+                throw new NotFoundException("Unable to find requested ref.");
+              }
+
+              throw new NotFoundException(
+                  String.format("[%d] object(s) missing in table read [%s]. \n\nObjects expected: %s\n\nObjects Received: %s",
```

Comment:
I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [FS: Format string should use %n rather than n (VA_FORMAT_STRING_USES_NEWLINE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#fs-format-string-should-use-n-rather-than-n-va-format-string-uses-newline).
This format string includes a newline character (\n). In format strings, it is generally  preferable to use %n, which will produce the platform-specific line separator.